### ISSUE TITLE
Add an option for spectators to toggle player colors

### ISSUE THREAD

**Issue by ghost** [OpenRA/OpenRA#15027]
An option that enables spectators to toggle player colors would help in many situations where it can be hard to distinguish between players. Although this request might be similar to #7497, it aims only towards improving the experience for spectators, casters and viewers. 

Why is it needed?

The existing function to avoid similar player colors does not ensure that only non-similar colors are used. It doesn't work for me personally because of colorblindness (Protanopia) but I heard from other players that they sometimes have a hard time to tell which unit belongs to whom too. 

Who would it help?

Having an option for spectators to toggle player colors would help spectators, casters and their viewers if there is a combination of similar colors. Players wouldn't be affected and kept their freedom to play with the color they like. 

Why not change the player colors and improve the function that should avoid similar colors?

Players should be free to pick the color they like. If similar colors are no problem for them, there is not a problem that needs to be solved in-game. An approach to improve color accessibility should incorporate colorblindness support which may be hard to implement properly given the different types of colorblindness. 

What would the toggle do?

It could either apply a standard color setup (red/blue in 1v1 is probably the most accessible setup) or calculate the complementary colors based on (a) given player color(s). Either way it should return colors that are easy to distinguish (at best for color blind people too, which could be up to 8% of OpenRA's playerbase - see https://www.smashingmagazine.com/2016/06/improving-color-accessibility-for-color-blind-users/).

Additional information regarding colorblindness fyi:
[Are you colorblind?](http://enchroma.com/test-150401.2/#&ui-state=dialog
)
[Accessibility of data visualization](https://www.dundas.com/support/blog/visualizing-for-the-color-blind
)
[Best practices in web design](https://www.usability.gov/get-involved/blog/2010/02/color-blindness.html)

And on a personal note regarding colorblindness: Whenever you visualize data make sure to use safe color combinations or combine them with graphical patterns.

**Comment by gdavegdave** [comment]
Yep, great idea. It would improve accessibility for people with colour blindness and it would help when player colours happen to be too similar to distinguish easily. 

To that end, it should be an option for both players and spectators. Perhaps a set list of colours could be selected based on avoiding issues with colour blindness (starting with the most common forms) and being easiest to distinguish in general. Then maybe the feature could be enabled with a single key press (or a click in the display settings). 

Displaying 'Player Stance Colours' is already an option, but that doesn't answer the need fully (e.g. to be able to tell your units apart from an ally) and it's less aesthetically pleasing to have colour bars.

**Comment by Smittytron** [comment]
Would also seeing radar vs tan players on ore and violet on gems. I believe SC2 has a button that toggles colors around in game.

**Comment by pchote** [comment]
Are there any global color-space modifying algorithms that could be applied here?
It would be much easier (and, if possible, more comprehensive) to apply this at the palette level so that it applies to *all* colors.

**Comment by lawando** [comment]
@pchote Adjusting colors for colorblind people is quite complex and I don't think that too much time should be invested in this. So far I only saw AAA games having full colorblindness support. There are algorithms for this as described in this scientific paper: http://www.cs.bham.ac.uk/~wbl/biblio/gecco2008/docs/p1065.pdf

Regarding high contrast colors:
I found this answer on SO that describes several different approaches for how to automatically generate N “distinct” colors: https://stackoverflow.com/a/4382138. Unfortunately I am currently not near my university so I can not access any of the pay-walled papers that contain the algorithms. The answer also includes lists of high contrast colors. Not sure if that is what you asked for. I can try to get access to the papers if you think it is helpful.

**Comment by ubitux** [comment]
I'd like to add one more use case in addition to the already mentioned color similarities and color blindness : observing a game with players who have the exact same color as the UI. I got confronted several times to that issue (admittedly with the same player), where at least the graph tabs in observer mode become invisible.

I believe the core of the issue lies into what you would expose to the user (a color shuffling button you would spam until it fits your need? selecting every player color individually? would it be available to players in-game as well as observers?)

On the other hand, choosing colors is not that hard of a problem I think. I usually go with HLS (Hue Lightness Saturation) with `light=0.6`, `saturation=1.0`, and play only with the `hue`.

Typical example:

```python
import sys, colorsys

n = int(sys.argv[1])
for i in range(n):
    hue = i / float(n)
    clr = colorsys.hls_to_rgb(hue, 0.6, 1.0)
    print('#' + ''.join('%02X' % int(c * 255) for c in clr))
```

```
% python ~/tmp/color.py 10
#FF3232
#FFAD32
#D6FF32
#5BFF32
#32FF84
#32FEFF
#3284FF
#5B32FF
#D632FF
#FF32AD
```

Basically, this:

![2019-04-06-204839_247x31_scrot](https://user-images.githubusercontent.com/34467/55673818-647cd580-58ad-11e9-8d2c-f0f34263178b.png)

**Comment by pchote** [comment]
> observing a game with players who have the exact same color as the UI. 

While it doesn't invalidate the request for this feature, the color validator really should be blocking these as invalid colours in the first place.  The validator seems to have regressed seriously some time ago...

**Comment by Mailaender** [comment]
Duplicate of https://github.com/OpenRA/OpenRA/issues/2462
